logic/dungeon5.py

```diff
@@ -5,82 +5,176 @@
-        m_stalfos_drop = Location(dungeon=5).add(HookshotDrop()).connect(third_arena, AND(FEATHER, SWORD)) # can reach fourth arena from entrance with feather and sword
+        # connections
+        # entrance / ms1
+        entrance.connect(entrance_ledge, HOOKSHOT, back=False)
+        entrance_ledge.connect(entrance_ledge_chest1, back=False)
+        entrance.connect(before_a_passage, AND(r.enemy_requirements["KEESE"], r.enemy_requirements["IRON_MASK"]), back=r.enemy_requirements["IRON_MASK"])
+        entrance.connect(crystal_puddle_room, FOUND(KEY5, 1))
+        before_a_passage.connect(before_a_passage_chest2, back=False)
+        crystal_puddle_room.connect((messy_room, statue_room), r.enemy_requirements["IRON_MASK"], back=False)
+        crystal_puddle_room.connect(ms_1_room, AND(r.enemy_requirements["STALFOS_AGGRESSIVE"], r.enemy_requirements["STALFOS_EVASIVE"]), back=False)
+        messy_room.connect(messy_room_chest3, r.enemy_requirements["IRON_MASK"], back=False)
+        ms_1_room.connect((ms_1_victory, crystal_puddle_room, before_d_passage), r.enemy_requirements["MASTER_STALFOS"], back=False)
+        before_d_passage.connect(ms_1_room, back=False)
+        # crossroads
+        before_d_passage.connect(before_d_passage_chest5, back=False)
+        before_d_passage.connect((north_crossroads, after_d_passage), FEATHER)
+        before_d_passage.connect((west_crossroads, ms_2_room))
+        west_crossroads.connect(west_crossroads_zol_clear, r.enemy_requirements["HIDING_ZOL"], back=False)
+        ms_2_room.connect(ms_2_victory, r.enemy_requirements["MASTER_STALFOS"], back=False)
+        north_crossroads.connect(single_tile_ledge, HOOKSHOT, back=False)
+        north_crossroads.connect(single_tile_ledge, FEATHER)
+        single_tile_ledge.connect(middle_ledge, AND(FEATHER, PEGASUS_BOOTS))
+        single_tile_ledge.connect(middle_ledge, HOOKSHOT, back=False)
+        middle_ledge.connect(middle_ledge_chest6, back=False)
+        middle_ledge.connect(east_ledge, HOOKSHOT)
+        east_ledge.connect(east_ledge_chest8, back=False)
+        middle_ledge.connect(north_ledge, HOOKSHOT, back=False)
+        north_ledge.connect(north_ledge_chest_7, back=False)
+        north_ledge.connect(north_crossroads, HOOKSHOT)
+        north_crossroads.connect(ms_3_room, AND("D5_ZOL_CLEAR", r.enemy_requirements["HIDING_ZOL"]), back=False)
+        ms_3_room.connect(ms_3_victory, AND("MS1_KILL", "MS2_KILL", r.enemy_requirements["MASTER_STALFOS"]), back=False)
+        ms_3_room.connect(pot_locked_room, POWER_BRACELET, back=False)
+        pot_locked_room.connect(pot_locked_room_chest9, AND(r.enemy_requirements["STALFOS_AGGRESSIVE"], r.enemy_requirements["STALFOS_EVASIVE"]), back=False) #TODO: remove kill enemy requirements, it's just here to make logic match stable
+        
+        # ms4
+        before_a_passage.connect(after_a_passage, FEATHER)
+        after_a_passage.connect(crystal_room, SWORD)

```

Comment:
can put `back=False` on connection to crystal room as well to be consistent I believe?